File: src/bun.js/bindings/DeferredError.zig
// Error's cannot be created off of the main thread. So we use this to store the
// information until its ready to be materialized later.
pub const DeferredError = struct {
    kind: Kind,
    code: jsc.Node.ErrorCode,
    msg: bun.String,

    pub const Kind = enum { plainerror, typeerror, rangeerror };

    pub fn from(kind: Kind, code: jsc.Node.ErrorCode, comptime fmt: [:0]const u8, args: anytype) DeferredError {
        return .{
            .kind = kind,
            .code = code,
            .msg = bun.handleOom(bun.String.createFormat(fmt, args)),
        };
    }

    pub fn toError(this: *const DeferredError, globalThis: *JSGlobalObject) JSValue {
        const err = switch (this.kind) {
            .plainerror => this.msg.toErrorInstance(globalThis),
            .typeerror => this.msg.toTypeErrorInstance(globalThis),
            .rangeerror => this.msg.toRangeErrorInstance(globalThis),
        };
        err.put(globalThis, ZigString.static("code"), ZigString.init(@tagName(this.code)).toJS(globalThis));
        return err;
    }
};

const bun = @import("bun");
const ZigString = @import("./ZigString.zig").ZigString;

const jsc = bun.jsc;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;

File: src/jsc/DeferredError.rs
use crate::node::ErrorCode;
use crate::{JSGlobalObject, JSValue, StringJsc, ZigStringJsc};
use bun_core::{String as BunString, ZigString};

// Error's cannot be created off of the main thread. So we use this to store the
// information until its ready to be materialized later.
pub struct DeferredError {
    pub kind: Kind,
    pub code: ErrorCode,
    pub msg: BunString,
}

#[derive(Copy, Clone, Eq, PartialEq)]
pub enum Kind {
    Plainerror,
    Typeerror,
    Rangeerror,
}

impl DeferredError {
    pub fn from(kind: Kind, code: ErrorCode, args: core::fmt::Arguments<'_>) -> DeferredError {
        DeferredError {
            kind,
            code,
            // bun.handleOom dropped: Rust allocation aborts on OOM by default.
            msg: BunString::create_format(args),
        }
    }

    pub fn to_error(&self, global: &JSGlobalObject) -> JSValue {
        let err = match self.kind {
            Kind::Plainerror => self.msg.to_error_instance(global),
            Kind::Typeerror => self.msg.to_type_error_instance(global),
            Kind::Rangeerror => self.msg.to_range_error_instance(global),
        };
        err.put(
            global,
            ZigString::static_(b"code"),
            ZigString::init(<&'static str>::from(self.code).as_bytes()).to_js(global),
        );
        err
    }
}

// ported from: src/jsc/DeferredError.zig
